### ISSUE TITLE
format.hms() accepting argument format

### ISSUE THREAD

**Issue by retodomax** [tidyverse/hms#125]
There seems to be a `format()` S3 method for class 'hms'. However, it just returns the input as character. I would have expected that it takes the argument `format` and formats the output like an object of class `POSIXct`

``` r
library(hms)
t <- Sys.time()
format(t, "%H")
#> [1] "17"

t2 <- as_hms(t)
format(t2, "%H")
#> [1] "17:12:38.169816"
```

<sup>Created on 2024-11-05 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by krlmlr** [comment]
Thanks. The hms type inherits from difftime, that type doesn't seem to support the `format()` syntax you're suggesting either:

``` r
d <- Sys.time() - Sys.time()
d
#> Time difference of -9.536743e-07 secs
format(d, "%H")
#> Error in prettyNum(.Internal(format(x, trim, digits, nsmall, width, 3L, : invalid 'trim' argument
```

<sup>Created on 2024-11-10 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Here's what I think happens: `format()` is forwarded to `format.default()`, and the `"%H"` is matched to the `trim` argument.

``` r
format.difftime
#> function (x, ...) 
#> {
#>     if (length(x)) 
#>         paste(format(unclass(x), ...), units(x))
#>     else character()
#> }
#> <bytecode: 0x10527d070>
#> <environment: namespace:base>
formals(format.default)
#> $x
#> 
#> 
#> $trim
#> [1] FALSE
#> 
#> $digits
#> NULL
#> 
#> $nsmall
#> [1] 0
#> 
#> $justify
#> c("left", "right", "centre", "none")
#> 
#> $width
#> NULL
#> 
#> $na.encode
#> [1] TRUE
#> 
#> $scientific
#> [1] NA
#> 
#> $big.mark
#> [1] ""
#> 
#> $big.interval
#> [1] 3
#> 
#> $small.mark
#> [1] ""
#> 
#> $small.interval
#> [1] 5
#> 
#> $decimal.mark
#> getOption("OutDec")
#> 
#> $zero.print
#> NULL
#> 
#> $drop0trailing
#> [1] FALSE
#> 
#> $...
```

<sup>Created on 2024-11-10 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by retodomax** [comment]
Thanks for the reply!
I guess, ideal would be, if this feature would be implemented in base R but hard to say if they want it there...
If you're interested, I could try to write a method implementing the feature for package `hms` (where in my opinion it would make sense to be able to extract/format specific time components (hours, minutes, ...) as string.

**Comment by retodomax** [comment]
Yes, so far, I also transformed the `hms` object to a `POSIXct` object to apply `format.POSIXct()`. This is useful for example to remove the seconds in a ggplot time axis:

```
library(tidyverse)
set.seed(1)
tibble(hour = seq(0,24,0.02),
       time = hms::hms(hours = hour),
       y = sin(hour/24*(2*pi)) + rnorm(length(hour), sd = 0.2)) %>% 
  ggplot(aes(x = time, y = y)) +
  geom_point() +
  geom_smooth() +
  scale_x_time(breaks = hms::hms(hours = c(0, 6, 12, 18, 24)),
               labels = \(x) format(x = as.POSIXct(x), "%H:%M"))
```

If there would be a `format.hms()` method with argument `format`, then this would be even easier : )

**Comment by moodymudskipper** [comment]
`format.hms()` formats object in a consistant "HH:MM:SS.ssssss" format with up to 6 decimals: 

``` r
hms::hms(12345)
#> 03:25:45
hms::hms(12345.12345678)
#> 03:25:45.123457
```

This format prints similarly to the POSIXct format, but doesn't offer the same flexibility, the digits arg is also ignored for hms, though it works with difftime.

``` r
t <- Sys.time()
format(t)
#> [1] "2025-03-21 13:22:19"
format(t, format = "%H:%M:%S")
#> [1] "13:22:19"
```

A way to support this, suggested in the thread, is to convert the hms object to a POSIXct object and then use the format method, however the hours would be limited to 24, which will confuse users, and formats for years, weeks, months will be available and make little sense.

Providing a glue specification like would be more flexible. The default format could be `"{.Hours}:{.minutes}:{.seconds_dec}"` where `.Hours` (capitalised) is total hours, `.minutes` (not capitalised) means minutes within hour, and `.seconds_dec` means decimal seconds within the minute, on which the `digits` arg applies. We could also have `.days`, `.weeks` and their variants.

**Comment by moodymudskipper** [comment]
My understanding was that hms returns durations, not intervals, so those leap seconds considerations wouldn't apply.
We can also make digits work, incl negative digits, that's less flexible but that would handle OP's case and hours > 24.

`%OS` prints fractional seconds but only if `options(digits.secs =)` is set.

**Comment by schochastics** [comment]
```
Specific to R is %OSn, which for output gives the seconds truncated to exactly 0 <= n <= 6 decimal places. If %OS is not followed by a digit, it uses up to digits decimals (only as many as necessary, i.e., not producing extra trailing zeros. The default digits = NULL corresponds to digits = 0. 
```

Would you support `%OSn` and digits? I think if formats are added to `format_hms()` it should probably quite simple. Something like `%H`, `%M` and `%S` and a separate digits argument?

**Comment by krlmlr** [comment]
@retodomax: What are the typical format strings you would use? Can we get away with supporting delimiters instead?

**Comment by krlmlr** [comment]
How about we implement a simple state machine that iterates over `strsplit(format, "")[[1]]` and creates arguments that are later passed on to `paste0()` ? The arguments would be characters or calls to `format_two_digits()` and such.

**Comment by moodymudskipper** [comment]
I had a quick look and none of 'hms', 'lubridate', or 'clock' seem to have a way of formatting a duration to say for instance "There are 90 minutes left". IMHO this is a missed opportunity.
I'm afraid that in the name of simplicity we're building a not only limited but also surprising solution:
* Supporting only a subset of the POSIX standard specifications
* Not supporting things like `format = "---%H:%M:%S---"` 
* Either not supporting hours > 24,  or mapping `"%H"` to hours more than 24 : both surprising

**Comment by schochastics** [comment]
Not sure this is so surprising. As far as I understand, a hms object is a duration, not a date, so supporting %H >24 is not that surprising and hms also indicates that it doesnt suddenly switch to dhms. I think with Kirills suggestion, one can support custom formats. My failed PR had a starting point I think.
This is a quick and dirty way for custom formatting when only %H %M %S are allowed
``` r
library(hms)
format_time <- function(hours, minutes, seconds, format = "%H:%M:%S") {
  formatted <- format
  formatted <- gsub("%H", hours, formatted)
  formatted <- gsub("%M", minutes, formatted)
  formatted <- gsub("%S", seconds, formatted)
  return(formatted)
}

format_hms1 <- function(x, format="%H:%M:%S"){
  xx <- hms:::decompose(x)
  hours <- hms:::format_hours(xx$hours)
  minutes <- hms:::format_two_digits(xx$minute_of_hour)
  seconds <- hms:::format_two_digits(xx$second_of_minute)
  format_time(hours,minutes,seconds, format)
}

x <- hms(123456)
format_hms1(x)
#> [1] "34:17:36"
format_hms1(x, format = "-----%H:%M:%S-----")
#> [1] "-----34:17:36-----"
format_hms1(x, format = "There are %M minutes left")
#> [1] "There are 17 minutes left"
```

<sup>Created on 2025-03-28 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by moodymudskipper** [comment]
Iif we map  "%H" to more than 24h we are not following the POSIX standard, and that we support it partially (+optionally if we don't handle identically escaping "%", non existent "%FOO" codes, or other things we might forget) we are copying a standard to have something that looks and feels the same to the user, then change its behaviour, rounding angles to suit our corner cases. I feel it's not proper practice. We can document it extensively, but I think we're pegging squares in round holes.

You can indeed do `"there are 27 minutes left"`, but not `"there are 90 minutes left"`, and counting in seconds, minutes, or hours without fancy extraction is useful.
